### ISSUE TITLE
[BUG] Contig boundary errors in anvi-export-locus

### ISSUE THREAD

**Issue by philipwoods** [merenlab/anvio#1841]
## Short description of the problem

`anvi-export-locus` sometimes fails to identify when a requested locus crosses a contig boundary.

## anvi'o version

```
Anvi'o .......................................: hope (v7.1)

Profile database .............................: 38
Contigs database .............................: 20
Pan database .................................: 15
Genome data storage ..........................: 7
Auxiliary data storage .......................: 2
Structure database ...........................: 2
Metabolic modules database ...................: 2
tRNA-seq database ............................: 2
```

## System info

OS is Red Hat Enterprise Linux release 8.3 (Ootpa).
Anvi'o is installed as a conda environment.

## Detailed description of the issue

I ran `anvi-export-locus` on several genomes asking for 5 genes on each side of the target, and in several of them the target gene is near the end of a contig. In most of these, Anvio properly trimmed down the gene calls to just those on the correct contig, but in three of them it did not. This manifested in a few different ways down the line.

In two of the problem loci, a contig boundary was only within 5 genes on one side of the target. One of them had the target in the forward direction, and the other had the target in the reverse direction. In these, the eventual failure was creating a gene calls file with negative start and/or stop positions for the genes in the locus, which then caused an error.

In the third problem locus, the target was within 5 genes of a contig boundary on both sides. Anvio properly trimmed the gene calls on one side of the target, but not on the other side. This resulted in a gene calls file with genes from two contigs, and eventually there was an error in utils.get_most_likely_translation_frame because one of the sequences was too short (presumably one of the genes from the wrong contig, as the others all seem fine).

## Files to reproduce

Here is a Dropbox link with the databases, the console output I saw for each one, and the incorrect gene calls files they created. I also included instructions to reproduce.

https://www.dropbox.com/s/q0ht6r0pleo4yni/anvi-export-locus.zip?dl=0

**Comment by mschecht** [comment]
@philipwoods thanks for documenting this and providing files to reproduce the error, this looks like a bug. I will look into it on Monday and get back to you.

**Comment by mschecht** [comment]
Hi @philipwoods, apologies for the delay. I just proposed a solution in PR #1846 on the main branch and hopefully this will be solved soon. 

You're contigsDBs have highlighted a known limitation of `anvi-export-locus` as discussed here: [Using anvi-export-locus with external gene calls](https://merenlab.org/2019/10/17/export-locus/#introduction). However, I did not anticipate the problem to arise with other internal anvio gene-callers, e.g. `Transfer_RNAs`, `Ribosomal_RNA_23S`, used in conjunction with Prodigal. Long story short, this bug you found  is making `anvi-export-locus` a better program and thanks again for sending a reproducible example!

**Comment by mschecht** [comment]
@philipwoods if you use the [anvio development branch](https://merenlab.org/2016/06/26/installation-v2/#5-follow-the-active-development-youre-a-wizard-arry) I think your problem should be solved.

I was able to run `anvi-export-locus` on your contigsDBs and extract 6 loci based on your search term:
```bash
mkdir -p rnfC

for f in ANME-2c_sp._UBA261_GCA_002506015.1.db ANME-3_sp._Jaco_11668_bin3.db S4_bin26_refined.db;
do
    fname=$(basename $f .db | sed 's|-||' | sed 's|\.||g')
    echo $fname
    anvi-export-locus -c $f -s "RnfC" -n 5,5 -O "${fname}"_RnfC -W -o rnfC --debug 
done

$ ls rnfC/
ANME2c_sp_UBA261_GCA_0025060151_RnfC_0001-PROFILE                 S4_bin26_refined_RnfC_0001-PROFILE
ANME2c_sp_UBA261_GCA_0025060151_RnfC_0001.db                      S4_bin26_refined_RnfC_0001.db
ANME2c_sp_UBA261_GCA_0025060151_RnfC_0001.fa                      S4_bin26_refined_RnfC_0001.fa
ANME2c_sp_UBA261_GCA_0025060151_RnfC_0001_external_gene_calls.txt S4_bin26_refined_RnfC_0001_external_gene_calls.txt
ANME2c_sp_UBA261_GCA_0025060151_RnfC_0001_sequence.fa             S4_bin26_refined_RnfC_0001_sequence.fa
ANME3_sp_Jaco_11668_bin3_RnfC_0001-PROFILE                        S4_bin26_refined_RnfC_0002-PROFILE
ANME3_sp_Jaco_11668_bin3_RnfC_0001.db                             S4_bin26_refined_RnfC_0002.db
ANME3_sp_Jaco_11668_bin3_RnfC_0001.fa                             S4_bin26_refined_RnfC_0002.fa
ANME3_sp_Jaco_11668_bin3_RnfC_0001_external_gene_calls.txt        S4_bin26_refined_RnfC_0002_external_gene_calls.txt
ANME3_sp_Jaco_11668_bin3_RnfC_0001_sequence.fa                    S4_bin26_refined_RnfC_0002_sequence.fa
ANME3_sp_Jaco_11668_bin3_RnfC_0002-PROFILE                        S4_bin26_refined_RnfC_0003-PROFILE
ANME3_sp_Jaco_11668_bin3_RnfC_0002.db                             S4_bin26_refined_RnfC_0003.db
ANME3_sp_Jaco_11668_bin3_RnfC_0002.fa                             S4_bin26_refined_RnfC_0003.fa
ANME3_sp_Jaco_11668_bin3_RnfC_0002_external_gene_calls.txt        S4_bin26_refined_RnfC_0003_external_gene_calls.txt
ANME3_sp_Jaco_11668_bin3_RnfC_0002_sequence.fa                    S4_bin26_refined_RnfC_0003_sequence.fa
```

Please let me know if it's working for you when you have a chance.

**Comment by philipwoods** [comment]
Sorry for the delay! That does seem to have resolved the issue. However, I have noticed a new one. When I used this new version with a different locus query, I occasionally got the following error:
```
Traceback (most recent call last):
  File "/export/data1/sw/anaconda3-2019.07/envs/anvio-dev/anvio/bin/anvi-export-locus", line 102, in <module>
    locus_splitter.process()
  File "/export/data1/sw/anaconda3-2019.07/envs/anvio-dev/anvio/anvio/splitter.py", line 885, in process
    self.export_locus(output_path_prefix, gene_callers_id)
  File "/export/data1/sw/anaconda3-2019.07/envs/anvio-dev/anvio/anvio/splitter.py", line 1074, in export_locus
    self.run.info("Locus gene call start/stops excess (nts)", excess)
UnboundLocalError: local variable 'excess' referenced before assignment
```
I'm not sure if this is related to the previous issue. Should I open a new bug report or just continue in this thread?

**Comment by mschecht** [comment]
Bummer, I think it is related! Can you send a reproducible example where you get that error?

**Comment by philipwoods** [comment]
Here are the two databases and the command I used with each of them. Thanks!
https://www.dropbox.com/s/vp790gxyw3zpsa8/locus-error.zip?dl=0

**Comment by mschecht** [comment]
Thanks so much, I'll try to address this tomorrow.

**Comment by mschecht** [comment]
@philipwoods sorry for the delay, this bug required some redesigning that took longer than anticipated. I addressed this new issue in PR #1852 

Thanks again for sending reproducible errors and please let me know if it's working on your end!

**Comment by philipwoods** [comment]
That seems to have fixed everything! Thank you so much!
